flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamInputProcessor.java

```diff
@@ -18,213 +18,17 @@
-	}
-
-	private void processElement(StreamElement recordOrMark, int channel) throws Exception {
-		if (recordOrMark.isRecord()) {
-			// now we can do the actual processing
-			StreamRecord<IN> record = recordOrMark.asRecord();
-			synchronized (lock) {
-				numRecordsIn.inc();
-				streamOperator.setKeyContextElement1(record);
-				streamOperator.processElement(record);
-			}
-		}
-		else if (recordOrMark.isWatermark()) {
-			// handle watermark
-			statusWatermarkValve.inputWatermark(recordOrMark.asWatermark(), channel);
-		} else if (recordOrMark.isStreamStatus()) {
-			// handle stream status
-			statusWatermarkValve.inputStreamStatus(recordOrMark.asStreamStatus(), channel);
-		} else if (recordOrMark.isLatencyMarker()) {
-			// handle latency marker
-			synchronized (lock) {
-				streamOperator.processLatencyMarker(recordOrMark.asLatencyMarker());
-			}
-		} else {
-			throw new UnsupportedOperationException("Unknown type of StreamElement");
-		}
-	}
-
-	private void initializeNumRecordsIn() {
-		if (numRecordsIn == null) {
-			try {
-				numRecordsIn = ((OperatorMetricGroup) streamOperator.getMetricGroup()).getIOMetricGroup().getNumRecordsInCounter();
-			} catch (Exception e) {
-				LOG.warn("An exception occurred during the metrics setup.", e);
-				numRecordsIn = new SimpleCounter();
-			}
-		}
-	}
-
-	private boolean checkFinished() throws Exception {
-		boolean isFinished = input.isFinished();
-		if (isFinished) {
-			synchronized (lock) {
-				operatorChain.endInput(1);
-			}
-		}
-		return isFinished;
-	}
-
-	public void cleanup() throws Exception {
-		input.close();
-	}
-
-	private class ForwardingValveOutputHandler implements StatusWatermarkValve.ValveOutputHandler {
-		private final OneInputStreamOperator<IN, ?> operator;
-		private final Object lock;
-
-		private ForwardingValveOutputHandler(final OneInputStreamOperator<IN, ?> operator, final Object lock) {
-			this.operator = checkNotNull(operator);
-			this.lock = checkNotNull(lock);
-		}
-
-		@Override
-		public void handleWatermark(Watermark watermark) {
-			try {
-				synchronized (lock) {
-					watermarkGauge.setCurrentWatermark(watermark.getTimestamp());
-					operator.processWatermark(watermark);
-				}
-			} catch (Exception e) {
-				throw new RuntimeException("Exception occurred while processing valve output watermark: ", e);
-			}
-		}
-
-		@SuppressWarnings("unchecked")
-		@Override
-		public void handleStreamStatus(StreamStatus streamStatus) {
-			try {
-				synchronized (lock) {
-					streamStatusMaintainer.toggleStreamStatus(streamStatus);
-				}
-			} catch (Exception e) {
-				throw new RuntimeException("Exception occurred while processing valve output stream status: ", e);
-			}
-		}
-	}
+public interface StreamInputProcessor extends AvailabilityListener, Closeable {
+	/**
+	 * @return false if probably there are not more records available and caller should check
+	 * {@link #isAvailable()}. return true if {@link #isAvailable()} check can be safely omitted.
```

Comment:
I'd rephrase this paragraph. Maybe something like:
```
Return true if more records can be processed.
False, if there are no more records available at the moment and the caller should check {@link #isFinished()} and/or {@link #isAvailable()}.
```

Reply:
I've rephrased your rephrased version to:
```
	/**
	 * @return true if {@link StreamInputProcessor} estimates that more records can be processed
	 * immediately. Otherwise false, which means that there are no more records available at the
	 * moment and the caller should check {@link #isFinished()} and/or {@link #isAvailable()}.
	 */
```